From 41dc22aed72a8b0d523a5a2a7bc4dc8fc694e753 Mon Sep 17 00:00:00 2001
From: Weilong Chen <chenweilong@huawei.com>
Date: Tue, 11 Apr 2023 17:01:06 +0800
Subject: [PATCH 1/2] iommu/arm-smmu-v3: Provide a module parameter for
 changing smmu queue length

Offering: HULK
ascend inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I7VD6D

--------------------------------

The default queue length of smmu is relatively large. When there
are a large number of smmu in the system, it will consume a lot
of memory. Provide configuration parameters to control and reduce
the memory floor noise. Use CONFIG_ASCEND_SMMU_QUEUE_SIZE as
compilation control.

Signed-off-by: Weilong Chen <chenweilong@huawei.com>
Signed-off-by: Guo Mengqi <guomengqi3@huawei.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 25 +++++++++++++++++++++
 1 file changed, 25 insertions(+)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 60ee198..60c7160 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -47,6 +47,13 @@ module_param(disable_msipolling, bool, 0444);
 MODULE_PARM_DESC(disable_msipolling,
 	"Disable MSI-based polling for CMD_SYNC completion.");
 
+#ifdef CONFIG_ASCEND_SMMU_QUEUE_SIZE
+static int qshift;
+module_param(qshift, int, 0444);
+MODULE_PARM_DESC(qshift,
+	"Set the length of cmdq priq, and eventq to control the memory usage of the smmu driver. Suggestion set to 10 arm-smmu-v3.qshift=10 (1024 entries). In order not to affect performance, add iommu.strict=0 to cmdline.");
+#endif
+
 #ifdef CONFIG_SMMU_BYPASS_DEV
 struct smmu_bypass_device {
 	unsigned short vendor;
@@ -5687,6 +5694,12 @@ static int arm_smmu_device_hw_probe(struct arm_smmu_device *smmu)
 	/* Queue sizes, capped to ensure natural alignment */
 	smmu->cmdq.q.llq.max_n_shift = min_t(u32, CMDQ_MAX_SZ_SHIFT,
 					     FIELD_GET(IDR1_CMDQS, reg));
+#ifdef CONFIG_ASCEND_SMMU_QUEUE_SIZE
+	/* Set cmdq size to qshift */
+	if (qshift > 0)
+		smmu->cmdq.q.llq.max_n_shift = min_t(u32, qshift,
+					     smmu->cmdq.q.llq.max_n_shift);
+#endif
 	if (smmu->cmdq.q.llq.max_n_shift <= ilog2(CMDQ_BATCH_ENTRIES)) {
 		/*
 		 * We don't support splitting up batches, so one batch of
@@ -5701,8 +5714,20 @@ static int arm_smmu_device_hw_probe(struct arm_smmu_device *smmu)
 
 	smmu->evtq.q.llq.max_n_shift = min_t(u32, EVTQ_MAX_SZ_SHIFT,
 					     FIELD_GET(IDR1_EVTQS, reg));
+#ifdef CONFIG_ASCEND_SMMU_QUEUE_SIZE
+	/* Set evtq size to qshift */
+	if (qshift > 0)
+		smmu->evtq.q.llq.max_n_shift = min_t(u32, qshift,
+					     smmu->evtq.q.llq.max_n_shift);
+#endif
 	smmu->priq.q.llq.max_n_shift = min_t(u32, PRIQ_MAX_SZ_SHIFT,
 					     FIELD_GET(IDR1_PRIQS, reg));
+#ifdef CONFIG_ASCEND_SMMU_QUEUE_SIZE
+	/* Set priq size to qshift */
+	if (qshift > 0)
+		smmu->priq.q.llq.max_n_shift = min_t(u32, qshift,
+					     smmu->priq.q.llq.max_n_shift);
+#endif
 
 	/* SID/SSID sizes */
 	smmu->ssid_bits = FIELD_GET(IDR1_SSIDSIZE, reg);
-- 
2.17.1

